drivers/misc/optee.c

```diff
@@ -145,10 +207,398 @@ static bool optee_safe_va_range(FAR void *va, size_t size)
+          list_entry++;
+          i = 0;
+        }
+    }
+
+  if (list_pa_p)
+    {
+      *list_pa_p = optee_va_to_pa(list) | shm_pgoff;
+    }
+
+  return list;
+}
+
+/****************************************************************************
+ * Name: optee_shm_sec_register
+ *
+ * Description:
+ *   Register specified shared memory entry with OP-TEE.
+ *
+ * Parameters:
+ *   priv  - The driver's private data structure
+ *   shme  - Pointer to shared memory entry to register. The entry, the
+ *           contained shared memory object, or the referenced shared buffer
+ *           cannot be NULL.
+ *
+ * Returned Values:
+ *   0 on success, negative error code otherwise.
+ *
+ ****************************************************************************/
+
+static int optee_shm_sec_register(FAR struct optee_priv_data *priv,
+                                  FAR struct optee_shm_entry *shme)
+{
+  FAR struct optee_shm_entry *msg_shme;
+  FAR struct optee_msg_arg *msg;
+  FAR void *page_list;
+  uintptr_t page_list_pa;
+  int ret = 0;
+
+  msg = optee_shm_alloc_msg(priv, 1, &msg_shme);
+  if (msg == NULL)
+    {
+      return -ENOMEM;
+    }
+
+  page_list = optee_shm_to_page_list(shme, &page_list_pa);
+  if (page_list == NULL)
+    {
+      ret = -ENOMEM;
+      goto errout_with_msg;
+    }
+
+  msg->cmd = OPTEE_MSG_CMD_REGISTER_SHM;
+  msg->params[0].attr = OPTEE_MSG_ATTR_TYPE_TMEM_OUTPUT |
+                        OPTEE_MSG_ATTR_NONCONTIG;
+  msg->params[0].u.tmem.buf_ptr = page_list_pa;
+  msg->params[0].u.tmem.shm_ref = shme->shm.addr;
+  msg->params[0].u.tmem.size = shme->shm.length;
+
+  if (optee_transport_call(priv, msg) || msg->ret)
+    {
+      ret = -EINVAL;
+    }
+
+  kmm_free(page_list);
+
+errout_with_msg:
+  optee_shm_free(msg_shme);
+
+  return ret;
+}
+
+/****************************************************************************
+ * Name: optee_shm_sec_unregister
+ *
+ * Description:
+ *   Unregister specified shared memory entry with OP-TEE.
+ *
+ * Parameters:
+ *   priv  - the driver's private data structure
+ *   shme  - Pointer to shared memory entry to unregister. The shared
+ *           memory entry must have been previously registered previously
+ *           with the OP-TEE and cannot be NULL.
+ *
+ * Returned Values:
+ *   0 on success, negative error code otherwise.
+ *
+ ****************************************************************************/
+
+static int optee_shm_sec_unregister(FAR struct optee_priv_data *priv,
+                                    FAR struct optee_shm_entry *shme)
+{
+  FAR struct optee_shm_entry *msg_shme;
+  FAR struct optee_msg_arg *msg;
+  int ret = 0;
+
+  msg = optee_shm_alloc_msg(priv, 1, &msg_shme);
+  if (msg == NULL)
+    {
+      return -ENOMEM;
+    }
+
+  msg->cmd = OPTEE_MSG_CMD_UNREGISTER_SHM;
+  msg->params[0].attr = OPTEE_MSG_ATTR_TYPE_RMEM_INPUT;
+  msg->params[0].u.rmem.shm_ref = shme->shm.addr;
+
+  if (optee_transport_call(priv, msg) || msg->ret)
```

Comment:
```
ret = optee_transport_call(priv, msg);
if (ret >= 0)
  {
    ret = optee_convert_error(msg->ret);
  }
```
let's add optee_convert_error to convert the optee error message to posix error code